buildSrc/src/main/java/com/gradle/publish/DownloadMavenArtifactsAndPublishToGradlePluginPortal.java

```diff
@@ -0,0 +1,273 @@
+    String baseUrl = mavenSourceRepositoryUrl.endsWith("/") ? mavenSourceRepositoryUrl : mavenSourceRepositoryUrl + "/";
+    String groupUrl = baseUrl + groupId.replace('.', '/');
+    String artifactDirectoryUrl = groupUrl + "/" + artifactId + "/" + version + "/";
+    LOGGER.info("Downloading artifacts to publish from  {}", artifactDirectoryUrl);
+    String artifactUrlPrefix = artifactDirectoryUrl + artifactId + "-" + version;
+
+    HttpClient client = HttpClient.newHttpClient();
+    downloadArtifact(client, artifactUrlPrefix + ".pom",
+      "build/publications/pluginMaven/pom-default.xml",
+      "build/publications/" + pluginConfigurationName + "PluginMarkerMaven/pom-default.xml");
+    downloadArtifact(client, artifactUrlPrefix + ".pom.asc",
+      "build/publications/pluginMaven/pom-default.xml.asc",
+      "build/publications/" + pluginConfigurationName + "PluginMarkerMaven/pom-default.xml.asc");
+    downloadArtifact(client, artifactUrlPrefix + ".jar", "build/libs/" + artifactId + "-" + version + ".jar");
+    downloadArtifact(client, artifactUrlPrefix + ".jar.asc", "build/libs/" + artifactId + "-" + version + ".jar.asc");
+    downloadArtifact(client, artifactUrlPrefix + "-sources.jar", "build/libs/" + artifactId + "-" + version + "-sources.jar");
+    downloadArtifact(client, artifactUrlPrefix + "-sources.jar.asc", "build/libs/" + artifactId + "-" + version + "-sources.jar.asc");
+    downloadArtifact(client, artifactUrlPrefix + ".module", "build/publications/pluginMaven/module.json");
+    downloadArtifact(client, artifactUrlPrefix + ".module.asc", "build/publications/pluginMaven/module.json.asc");
+    downloadArtifact(client, artifactUrlPrefix + "-javadoc.jar", "build/libs/" + artifactId + "-" + version + "-javadoc.jar");
+    downloadArtifact(client, artifactUrlPrefix + "-javadoc.jar.asc", "build/libs/" + artifactId + "-" + version + "-javadoc.jar.asc");
+  }
+
+  /**
+   * Validate that minimum publish content contains: pom, jar, sources, javadoc, module and their signatures.
+   */
+  private void validatePublishContent() {
+    LOGGER.info("ValidatePublishContent artifacts");
+    Map<PublishArtifact, File> artifacts = collectArtifacts();
+    artifacts.forEach((artifact, file) -> LOGGER.info("Publish artifact {} from file {}", artifact.encode(), file));
+
+    Set<String> actualArtifactTypes = artifacts.keySet().stream().map(PublishArtifact::getType).collect(Collectors.toSet());
+
+    Set<String> missingArtifactTypes = REQUIRED_ARTIFACT_TYPES.stream()
+      .filter(type -> !actualArtifactTypes.contains(type))
+      .collect(Collectors.toSet());
+
+    if (!missingArtifactTypes.isEmpty()) {
+      String missingArtifactTypesString = String.join(", ", missingArtifactTypes);
+      throw new GradleException("Missing required artifact types: " + missingArtifactTypesString);
+    }
+  }
+
+  public void downloadArtifact(HttpClient client, String artifactUrl, String... destinationPaths) throws InterruptedException {
```

Comment:
In which case are we using multiple destination paths? Can we just make it a single path?